platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/sync/protocol/SyncProtocol.java

```diff
@@ -0,0 +1,282 @@
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean shouldInitiate() {
+        syncMetrics.opportunityToInitiateSync();
+
+        // are there any reasons not to initiate?
+        if (!syncCooldownComplete()
+                || !peerAgnosticSyncChecks.shouldSync()
+                || fallenBehindManager.hasFallenBehind()) {
+            syncMetrics.updateDeclinedToInitiateSyncRatio(true);
+            return false;
+        }
+
+        // is there a reason to initiate?
+        if (peerNeededForFallenBehind() || criticalQuorum.isInCriticalQuorum(peerId.getId())) {
+            permit = permitProvider.tryAcquire();
+            final boolean isLockAcquired = permit.isLockAcquired();
+
+            if (isLockAcquired) {
+                syncMetrics.updateSyncPermitsAvailable(permitProvider.getNumAvailable());
+                syncMetrics.outgoingSyncRequestSent();
+            }
+
+            syncMetrics.updateDeclinedToInitiateSyncRatio(!isLockAcquired);
+
+            return isLockAcquired;
+        } else {
+            syncMetrics.updateDeclinedToInitiateSyncRatio(true);
+            return false;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean shouldAccept() {
+        syncMetrics.incomingSyncRequestReceived();
+
+        // are there any reasons not to accept?
+        if (!syncCooldownComplete()
+                || !peerAgnosticSyncChecks.shouldSync()
+                || fallenBehindManager.hasFallenBehind()) {
+            syncMetrics.updateRejectedSyncRatio(true);
+            return false;
+        }
+
+        permit = permitProvider.tryAcquire();
+        final boolean isLockAcquired = permit.isLockAcquired();
+
+        if (isLockAcquired) {
+            syncMetrics.updateSyncPermitsAvailable(permitProvider.getNumAvailable());
+            syncMetrics.acceptedSyncRequest();
+        }
+
+        syncMetrics.updateRejectedSyncRatio(!isLockAcquired);
+
+        return isLockAcquired;
+    }
+
+    /**
+     * Closes the existing permit, and resets the {@link #permit} member variable
+     */
+    private void closePermit() {
+        permit.close();
+        permit = MaybeLocked.NOT_ACQUIRED;
+
+        syncMetrics.updateSyncPermitsAvailable(permitProvider.getNumAvailable());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void initiateFailed() {
+        closePermit();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void acceptFailed() {
+        closePermit();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean acceptOnSimultaneousInitiate() {
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void runProtocol(@NonNull final Connection connection)
+            throws NetworkProtocolException, IOException, InterruptedException {
+
+        Objects.requireNonNull(connection);
```

Comment:
This connection will never be null, but if it is, this will cause the method to exit without closing the permit.

Reply:
Do you think I should remove the null check entirely, or just move it into the try block? I'm still trying to get a handle on how militant we want to be using the null checks

Reply:
I would just remove the check